storage/client_split_test.go

```diff
@@ -1574,3 +1575,107 @@ func writeRandomTimeSeriesDataToRange(
+func TestStoreSplitBeginTxnPushMetaIntentRace(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	splitKey := roachpb.RKey("a")
+
+	sCtx := storage.TestStoreContext()
+	sCtx.TestingKnobs.DisableSplitQueue = true
+	store, stopper, manual := createTestStoreWithContext(t, sCtx)
+	defer stopper.Stop()
+
+	// Advance the clock past the transaction cleanup expiration.
+	manual.Increment(storage.GetGCQueueTxnCleanupThreshold().Nanoseconds() + 1)
+
+	// First, create a split after addressing records.
+	args := adminSplitArgs(roachpb.KeyMin, keys.SystemPrefix)
+	if _, pErr := client.SendWrapped(rg1(store), nil, &args); pErr != nil {
+		t.Fatal(pErr)
+	}
+
+	startTime := timeutil.Now()
+	wroteMeta2 := make(chan struct{}, 1)
+	store.TestingKnobs().TestingCommandFilter = func(filterArgs storagebase.FilterArgs) *roachpb.Error {
+		if _, ok := filterArgs.Req.(*roachpb.BeginTransactionRequest); ok {
+			// Return a node unavailable error for BeginTransaction request
+			// in the event we haven't waited 20ms.
+			if timeutil.Now().Sub(startTime) < 20*time.Millisecond {
+				return roachpb.NewError(&roachpb.NodeUnavailableError{})
+			}
+			if log.V(1) {
+				log.Infof(context.TODO(), "allowing BeginTransaction to proceed after 20ms")
+			}
+		} else if filterArgs.Req.Method() == roachpb.Put &&
+			bytes.Equal(filterArgs.Req.Header().Key, keys.RangeMetaKey(splitKey)) {
+			select {
+			case wroteMeta2 <- struct{}{}:
+			default:
+			}
+		}
+		return nil
+	}
+
+	// Initiate split at splitKey in a goroutine.
+	doneSplit := make(chan *roachpb.Error, 1)
+	go func() {
+		if log.V(1) {
+			log.Infof(context.TODO(), "splitting at %s", splitKey)
+		}
+		args := adminSplitArgs(keys.SystemMax, splitKey.AsRawKey())
+		_, pErr := client.SendWrappedWith(store, nil, roachpb.Header{
+			RangeID: store.LookupReplica(splitKey, nil).RangeID,
+		}, &args)
+		doneSplit <- pErr
+	}()
+
+	// Wait for the write to the meta2 key to be initiated.
+	<-wroteMeta2
+
+	// Wait for 5ms to allow meta2 write to finish.
+	time.Sleep(5 * time.Millisecond)
+
+	// GC the replica containing the range descriptor records.
+	if err := store.ManualReplicaGC(store.LookupReplica(splitKey, nil)); err != nil {
+		// TODO(spencer): test fatal here after the gc queue correctly
+		// uses dist sender, because there is a chance the gc spans ranges.
+		//t.Fatal(err)
+		log.Errorf(context.TODO(), "failed gc queue: %s", err)
+	}
+
+	// Wait for the split to complete.
+	if pErr := <-doneSplit; pErr != nil {
+		t.Fatalf("failed split at %s: %s", splitKey, pErr)
+	}
+
+	// Now verify that the meta2/splitKey meta2 record is present; do this
+	// within a SucceedSoon in order to give the intents time to resolve.
+	util.SucceedsSoon(t, func() error {
+		val, intents, err := engine.MVCCGet(context.Background(), store.Engine(),
+			keys.RangeMetaKey(splitKey), hlc.MaxTimestamp, true, nil)
+		if err != nil {
+			return err
+		}
+		if val == nil {
+			t.Errorf("expected meta2 record for %s", keys.RangeMetaKey(splitKey))
```

Comment:
These are both test errors. The succeeds construct only runs as long as there are any errors getting the meta2 record due to intents.